core/navigation.js

```diff
@@ -425,43 +522,16 @@ Blockly.Navigation.findBestConnection = function(block, connection) {
  * intelligent guess about which connection to use to on the moving block.
  * @param {!Blockly.Block} block The block to move.
  * @param {Blockly.Connection} targetConnection The connection to connect to.
+ * @return {boolean} Whether the connection was successful.
  */
 Blockly.Navigation.insertBlock = function(block, targetConnection) {
   var bestConnection =
       Blockly.Navigation.findBestConnection(block, targetConnection);
-
-  if (bestConnection) {
-    if (targetConnection.type == Blockly.PREVIOUS_STATEMENT
-        || targetConnection.type == Blockly.OUTPUT_VALUE) {
-      block.positionNearConnection(bestConnection, targetConnection);
-    }
-    try {
-      targetConnection.connect(bestConnection);
-    }
-    catch (Error) {
-      // TODO: Is there anything else useful to do at this catch?
-      // Perhaps position the block near the target connection?
-      Blockly.Navigation.warn('The connection block is not the right type');
-    }
-  }
-};
-
-/**
- * Tries to connect the current location of the cursor and the insertion
- * connection.
- */
-Blockly.Navigation.insertBlockFromWs = function() {
-  var targetConnection = Blockly.Navigation.getInsertionConnection();
-  var sourceConnection = Blockly.Navigation.cursor_.getCurNode().getLocation();
-  if (targetConnection && sourceConnection) {
-    try {
-      sourceConnection.connect(targetConnection);
-    } catch (Error) {
-      Blockly.Navigation.warn('The connection block is not the right type');
-    }
+  if (bestConnection.isConnected() &&
+      !bestConnection.targetConnection.getSourceBlock().isShadow()) {
```

Comment:
nit: could we use besetConnection.targetBlock() instead or bestConnection.targetConnection.getSourceBlock()?